app/routes/datasets/queries.py

```diff
@@ -552,6 +563,102 @@ async def _add_geometry_filter(parsed_sql, geometry: Geometry):
     return parsed_sql
 
 
+async def _add_revision_filter(parsed_sql, dataset, version):
+    """Add revision filters to geostore table queries.
+
+    For queries of latest dataset versions, this involves excluding rows from revisions
+    that are deleted by subsequent revisions in the version's revision history.
+
+    For queries of intermediate revisions, this involves including all rows from
+    append revisions and excluding rows from deleted revisions in the revision's
+    history.
+    """
+
+    default_asset: AssetORM = await assets.get_default_asset(dataset, version)
+
+    if default_asset.asset_type == AssetType.revision:
+        source_version = default_asset.revision_history[0]["version"]
+        source_default_asset: AssetORM = await assets.get_default_asset(
+            dataset, source_version
+        )
+        revision_asset = default_asset
+        latest_revision = source_default_asset.latest_revision
+
+    else:
+        latest_revision = default_asset.latest_revision
+        try:
+            revision_asset: AssetORM = await assets.get_default_asset(
+                dataset, latest_revision
+            )
+        except RecordNotFoundError:
+            revision_asset = None
+
+    sql_where = parsed_sql[0]["RawStmt"]["stmt"]["SelectStmt"].get("whereClause", None)
+
+    if revision_asset is None:
+        revision_filters = []
+    elif revision_asset.version == latest_revision:
+        revision_filters = await _filter_by_revision_operation(revision_asset, "latest")
+    else:
+        revision_filters = await _filter_by_revision_operation(
+            revision_asset, "intermediate"
+        )
+
+    if not revision_filters:
+        return parsed_sql
+
+    if sql_where:
+        parsed_sql[0]["RawStmt"]["stmt"]["SelectStmt"]["whereClause"] = {
+            "BoolExpr": {"boolop": 0, "args": [*sql_where, *revision_filters]}
+        }
+    else:
+        parsed_sql[0]["RawStmt"]["stmt"]["SelectStmt"]["whereClause"] = {
+            "BoolExpr": {"boolop": 0, "args": revision_filters}
+        }
+
+    return parsed_sql
+
+
+async def _filter_by_revision_operation(revision_asset: AssetORM, filter_type: str):
+    delete_revisions = []
+    append_revisions = []
+    for revision in revision_asset.revision_history:
+        asset: AssetORM = await assets.get_default_asset(
+            revision["dataset"], revision["version"]
+        )
+        delete_revision = asset.creation_options.get("delete_version", None)
+
+        if delete_revision is not None:
+            delete_revisions.append(delete_revision)
+        else:
+            append_revisions.append(revision["version"])
+
+    if filter_type == "latest":
```

Comment:
I don't think you need this fork in logic based on a special keyword. I think both cases can latest and intermediate cases can apply the same logic. Pass both latest revision history and the current revision, then:

- Fill append_revisions with any append revisions that come after the current revision in the latest revision history (so this will be empty for the latest revision, but populated for intermediate revisions)
- Fill delete_revisions with all delete revisions
- Combine append_revisions and delete_revisions into a set to delete any duplicates
- Make your filter:

`WHERE gfw_version NOT IN ([revision for revision in combined_revisions])`

Reply:
That's neat, thanks!